src/renderer/utils/createStorage.ts

```diff
@@ -1,73 +1,210 @@
+        this.merge(value);
       }
+      this.initialized = true;
+    } catch (error) {
+      console.error(`[init]: ${error}`, this);
     }
+  }
 
-    return this.defaultValue;
+  isDefaultValue(value: T): boolean {
+    return isEqual(value, this.defaultValue);
   }
 
-  set(value: T) {
-    this.storage.setItem(this.key, JSON.stringify(value));
+  @action
+  private configure({ storage, observable }: StorageConfiguration<T> = this.options): this {
+    if (storage) this.storage = storage;
+    if (observable) this.configureObservable(observable);
 
     return this;
   }
 
-  merge(value: Partial<T>) {
-    const currentValue = this.get();
+  @action
+  private configureObservable(options: CreateObservableOptions = {}) {
+    this.data = observable.box<T>(this.data.get(), {
+      ...StorageHelper.defaultOptions.observable, // inherit default observability options
+      ...options,
+    });
+    this.data.observe(change => {
+      const { newValue, oldValue } = toJS(change, { recurseEverything: true });
+
+      this.onChange(newValue, oldValue);
+    });
+  }
 
-    return this.set(Object.assign(currentValue, value));
+  protected onChange(value: T, oldValue?: T) {
+    if (!this.initialized) return;
+
+    this.storage.onChange?.({ value, oldValue, key: this.key });
+    this.storage.setItem(this.key, value);
   }
 
-  clear() {
-    this.storage.removeItem(this.key);
+  load(): T {
+    return this.storage.getItem(this.key);
+  }
 
-    return this;
+  get(): T {
+    return this.data.get();
   }
 
-  getDefaultValue() {
-    return this.defaultValue;
+  set(value: T) {
+    this.data.set(value);
+  }
+
+  reset() {
+    this.set(this.defaultValue);
+  }
+
+  clear() {
+    this.data.set(null);
   }
 
-  restoreDefaultValue() {
-    return this.set(this.defaultValue);
+  merge(value: Partial<T> | ((draft: Draft<T>) => Partial<T> | void)) {
+    const updater = isFunction(value) ? value : (state: Draft<T>) => {
+      if (isObject(state)) return Object.assign(state, value);
+
+      return value;
+    };
+    const currentValue = this.toJS();
+    const nextValue = produce(currentValue, updater) as T;
+
+    this.set(nextValue);
+  }
+
+  toJS() {
+    return toJS(this.get(), { recurseEverything: true });
   }
 }
+
+export const localStorageAdapter: StorageAdapter<object> = {
+  getItem(key: string) {
+    return JSON.parse(localStorage.getItem(key));
+  },
+  setItem(key: string, value: any) {
+    localStorage.setItem(key, JSON.stringify(value));
+  },
+  removeItem(key: string) {
+    localStorage.removeItem(key);
+  }
+};
+
+// TODO: remove after merge https://github.com/lensapp/lens/pull/2269
+export const jsonFileSyncStorageAdapter: StorageAdapter<object> = {
+  cwd: path.resolve((app || remote.app).getPath("userData"), "lens-local-storage"),
+  data: observable.map<string, any>([], { deep: false }),
+  initialized: false,
+
+  get filePath() {
+    const clusterId = location.hostname.split(".").slice(-2, -1)[0];
```

Comment:
This should be gotten from `getHostedClusterId()`

Reply:
Yes, but no.. This is `renderer/utils/*` which should be with minimal dependencies, especially from `main` module.

Reply:
I guess I didn't realize that `getHostedClusterId()` was in `main`. That seems like a strange place to have it.

Reply:
Technically it's in `src/common/cluster-store.ts` but I consider it more as `main` stuff which should be avoided in `renderer`...

Reply:
How is it `main` then? It is used by `getHostedCluster()` all the time.